src/interrupt/kvm_irq/legacy_irq.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2019 Alibaba Cloud. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Manage virtual device's legacy interrupts based on Linux KVM framework.
+//!
```

Comment:
So this is an in-kernel KVM IRQ controller implementation. How would we support alternative legacy IRQ implementations like e.g. userspace IOAPIC?
I think that as part of the `InterruptManager` trait we should be able to dynamically register a constructor for each interrupt source type.
The idea would be for a userspace IOAPIC implementation to implement the `InterruptSourceGroup` trait and then for the VMM to decide if the KVM irq manager should use `LegacyIrq`'s constructor to create new legacy interrupt source groups or something else.

Reply:
Aha, missed the split-irqchip case. So we need to support two types of backends for legacy IRQ. 
May be we could support split-irqchip by:
1) Add an method to the IrqManager to register userspace irqchip driver.
2) Enhance the LegacyIrq to use userspace Irqchip if one is registered, otherwise use the in-kernel irqchip.
How about supporting split-irqchip after this PR has settle down?

Reply:
> Aha, missed the split-irqchip case. So we need to support two types of backends for legacy IRQ.
> May be we could support split-irqchip by:
> 
> 1. Add an method to the IrqManager to register userspace irqchip driver.

Yes.

> 2. Enhance the LegacyIrq to use userspace Irqchip if one is registered, otherwise use the in-kernel irqchip.

That makes sense.

>    How about supporting split-irqchip after this PR has settle down?

Sounds good to me.